stan/math/rev/core/var.hpp

```diff
@@ -63,185 +74,77 @@ class var {
-  // and unsigned integers, because long and unsigned long
-  // are still 32-bit
+  template <typename S, require_vari_convertible_t<S>* = nullptr>
+  var_value(vari_value<S>* vi)  // NOLINT
+      : vi_(vi) {}
 
   /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
+   * Construct a variable from a pointer to a variable implementation when.
+   *  the `value_type`s differ. This forces a copy of the vari
+   * @tparam S the type in the vari_value* that does not have the same type
+   *  as the `value_type` of this `var_value`.
+   * @param vi A `vari_value`.
    */
-  var(size_t x) : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
+  template <typename S, require_not_vari_convertible_t<S>* = nullptr>
+  var_value(vari_value<S>* vi) : vi_(new vari_type(*vi)) {}  // NOLINT
 
   /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
+   * Construct from a var_value whose underlying value_type differs in type
+   * to this class's value_type. This is for cases such as
+   * `var_value<double> a(5.0); var_value<long double> b(a);`
+   * @tparam S A type that is not the same as the underlying value type.
+   *  same as the underlying value_type.
+   * param x a `var_value` whose underlying vari_type can be dynamically cast
+   * to `this::vari_value<value_type>``.
    */
-  var(ptrdiff_t x) : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-#endif
-
-#ifdef BOOST_MATH_USE_FLOAT128
-
-  // this ctor is for later GCCs that have the __float128
-  // type enabled, because it gets enabled by boost
+  template <typename S, require_not_vari_convertible_t<S>* = nullptr>
+  var_value(const var_value<S>& x) : vi_(new vari_type(*x.vi_)) {}  // NOLINT
 
   /**
-   * Construct a variable from the specified arithmetic argument
-   * by constructing a new <code>vari</code> with the argument
-   * cast to <code>double</code>, and a zero adjoint.
-   *
-   * @param x Value of the variable.
+   * Constructor from `var_value` whose value_type is the same as this class's
+   * `value_type`. This is used in cases such as
+   * `var_value<double> a(4.0); var_value<int> b(a)` since the `value_type` for
+   *  a `var_value` with an integral type is a double.
    */
-  var(__float128 x) : vi_(new vari(static_cast<double>(x), false)) {}  // NOLINT
-
-#endif
+  template <typename S, require_vari_convertible_t<S>* = nullptr>
+  var_value(const var_value<S>& x) : vi_(x.vi_) {}  // NOLINT
 
   /**
    * Return the value of this variable.
    *
    * @return The value of this variable.
    */
-  inline double val() const { return vi_->val_; }
+  inline auto val() const { return vi_->val_; }
```

Comment:
auto on a const function will return a const reference?

Reply:
It will return back the type of `vi_->val_`. `auto&` on a const member function will return back a const reference though

https://godbolt.org/z/-bZs7A

Reply:
Should it be `auto&` then?

Reply:
Or I guess `const auto&` just for clarity? `val_` should be const.

Reply:
I'll make these const correct

Reply:
Cuz then `val()` will return a const& and `adj()` can return either a const& or just a &

Reply:
Do we even need a separate const option? I think leaving just the non-const one (returning a value) would be jsut fine.

Reply:
Eh, it matters whether we want to allow `a.adj() = y`. I think that's nice to have since we have the for dynamic matrices